src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.cs

```diff
@@ -4,133 +4,116 @@
+        // our cancellation token check will mean its a no-op.
+        _semaphore.Release();
+        _semaphore.Dispose();
 
-                var lcts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-                var updateTask = Task.Factory.StartNew(
-                    () => UpdateWorkspaceStateAsync(workspaceProject, projectSnapshot, lcts.Token),
-                    lcts.Token,
-                    TaskCreationOptions.None,
-                    TaskScheduler.Default).Unwrap();
-                updateItem = new UpdateItem(updateTask, lcts);
-                Updates[projectSnapshot.Key] = updateItem;
-            },
-            cancellationToken);
+        _blockBackgroundWorkStart?.Set();
     }
 
-    public void Dispose()
+    public void EnqueueUpdate(Project? workspaceProject, IProjectSnapshot projectSnapshot)
     {
-        if (_disposed)
+        if (_disposeTokenSource.IsCancellationRequested)
         {
             return;
         }
 
-        _disposed = true;
-
-        foreach (var update in Updates)
+        lock (_updates)
         {
-            if (!update.Value.Task.IsCompleted &&
-                !update.Value.Cts.IsCancellationRequested)
+            if (_updates.TryGetValue(projectSnapshot.Key, out var updateItem))
             {
-                update.Value.Cts.Cancel();
+                _logger.LogTrace($"Cancelling previously enqueued update for '{projectSnapshot.FilePath}'.");
+
+                updateItem.Dispose();
             }
+
+            _logger.LogTrace($"Enqueuing update for '{projectSnapshot.FilePath}'");
+
+            _updates[projectSnapshot.Key] = new UpdateItem(
+                token => UpdateWorkspaceStateAsync(workspaceProject, projectSnapshot, token),
+                _disposeTokenSource.Token);
         }
+    }
 
-        // Release before dispose to ensure we don't throw exceptions from the background thread trying to release
-        // while we're disposing. Multiple releases are fine, and if we release and it lets something passed the lock
-        // our cancellation token check will mean its a no-op.
-        _semaphore.Release();
-        _semaphore.Dispose();
+    public void CancelUpdates()
+    {
+        _logger.LogTrace($"Cancelling all previously enqueued updates.");
+
+        lock (_updates)
+        {
+            foreach (var (_, updateItem) in _updates)
+            {
+                updateItem.Dispose();
+            }
 
-        BlockBackgroundWorkStart?.Set();
+            _updates.Clear();
+        }
     }
 
     private async Task UpdateWorkspaceStateAsync(Project? workspaceProject, IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
     {
-        // We fire this up on a background thread so we could have been disposed already, and if so, waiting on our semaphore
-        // throws an exception.
-        if (_disposed)
+        if (_disposeTokenSource.IsCancellationRequested)
```

Comment:
Why does this check the `_disposeTokenSource` and not the passed in `cancellationToken`? Aren't they supposed to be linked?

Reply:
It only checks it because I was changing `_disposed` to `_disposeTokenSource`.

Honestly, I think this check can just be removed. The same token is already passed to the call to `Task.Run`, so it'd be a very rare situation where we get here and cancellation has been requested. I think it's fine to fall into the code below that calls `await _semaphore.WaitAsync(cancellationToken)`. That'll just return a cancelled task and the await will throw and hit the catch handler.

Reply:
I ended up removing that check and updated another one below.

Reply:
Hey friends, I think maybe we should leave this one :)